File: aun/aun-sdk-core/js/src/crypto.ts
// ── 密码学提供者（浏览器 SubtleCrypto 实现）──────────────
// 所有操作均为异步（SubtleCrypto API 要求）

/** Base64 编码（浏览器兼容） */
function uint8ToBase64(bytes: Uint8Array): string {
  let binary = '';
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

/** Base64 解码（浏览器兼容） */
function base64ToUint8(b64: string): Uint8Array {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

/** 将 ArrayBuffer 转为 PEM 格式字符串 */
function arrayBufferToPem(buffer: ArrayBuffer, label: string): string {
  const b64 = uint8ToBase64(new Uint8Array(buffer));
  const lines: string[] = [];
  for (let i = 0; i < b64.length; i += 64) {
    lines.push(b64.slice(i, i + 64));
  }
  return `-----BEGIN ${label}-----\n${lines.join('\n')}\n-----END ${label}-----`;
}

/** 从 PEM 字符串中提取 DER 数据 */
function pemToArrayBuffer(pem: string): ArrayBuffer {
  const b64 = pem
    .replace(/-----BEGIN [^-]+-----/, '')
    .replace(/-----END [^-]+-----/, '')
    .replace(/\s/g, '');
  const bytes = base64ToUint8(b64);
  return bytes.buffer;
}

/** 密码学提供者 — 使用浏览器 SubtleCrypto 实现 P-256 ECDSA */
export class CryptoProvider {
  static readonly curveName = 'P-256';

  /**
   * 生成 P-256 ECDSA 密钥对（异步）。
   * 返回 { private_key_pem, public_key_der_b64, curve }
   */
  async generateIdentity(): Promise<{
    private_key_pem: string;
    public_key_der_b64: string;
    curve: string;
  }> {
    const keyPair = await crypto.subtle.generateKey(
      { name: 'ECDSA', namedCurve: 'P-256' },
      true, // extractable
      ['sign', 'verify'],
    );

    // 导出 PKCS8 私钥
    const pkcs8 = await crypto.subtle.exportKey('pkcs8', keyPair.privateKey);
    const privateKeyPem = arrayBufferToPem(pkcs8, 'PRIVATE KEY');

    // 导出 SPKI 公钥（DER 格式，base64 编码）
    const spki = await crypto.subtle.exportKey('spki', keyPair.publicKey);
    const publicKeyDerB64 = uint8ToBase64(new Uint8Array(spki));

    return {
      private_key_pem: privateKeyPem,
      public_key_der_b64: publicKeyDerB64,
      curve: CryptoProvider.curveName,
    };
  }

  /**
   * 签名登录 nonce（异步）。
   * 返回 [signatureBase64, clientTime]
   *
   * 签名数据格式: "{nonce}:{clientTime}"
   * 使用 ECDSA + SHA-256
   */
  async signLoginNonce(
    privateKeyPem: string,
    nonce: string,
    clientTime?: string,
  ): Promise<[string, string]> {
    const usedTime = clientTime ?? String(Date.now() / 1000);
    const signData = new TextEncoder().encode(`${nonce}:${usedTime}`);

    // 导入 PEM 私钥
    const pkcs8 = pemToArrayBuffer(privateKeyPem);
    const cryptoKey = await crypto.subtle.importKey(
      'pkcs8',
      pkcs8,
      { name: 'ECDSA', namedCurve: 'P-256' },
      false,
      ['sign'],
    );

    // 签名（SubtleCrypto 返回 IEEE P1363 格式，与 Python cryptography DER 不同）
    const signature = await crypto.subtle.sign(
      { name: 'ECDSA', hash: 'SHA-256' },
      cryptoKey,
      signData,
    );

    // 将 P1363 格式转为 DER 格式（与 Python SDK 兼容）
    const derSignature = p1363ToDer(new Uint8Array(signature));
    return [uint8ToBase64(derSignature), usedTime];
  }

  /** 生成客户端 nonce（12 字节随机数的 base64 编码） */
  newClientNonce(): string {
    const bytes = new Uint8Array(12);
    crypto.getRandomValues(bytes);
    return uint8ToBase64(bytes);
  }
}

// ── 签名格式转换工具 ──────────────────────────────────────

/**
 * 将 IEEE P1363 格式签名转为 DER 格式。
 * SubtleCrypto 输出 P1363（r || s 固定长度拼接），
 * Python cryptography 使用 DER（ASN.1 编码）。
 */
function p1363ToDer(p1363: Uint8Array): Uint8Array {
  const halfLen = p1363.length / 2;
  const r = trimLeadingZeros(p1363.slice(0, halfLen));
  const s = trimLeadingZeros(p1363.slice(halfLen));

  // 若最高位为 1 需要前导 0x00（ASN.1 有符号整数）
  const rPadded = r[0] & 0x80 ? new Uint8Array([0, ...r]) : r;
  const sPadded = s[0] & 0x80 ? new Uint8Array([0, ...s]) : s;

  const rTlv = new Uint8Array([0x02, rPadded.length, ...rPadded]);
  const sTlv = new Uint8Array([0x02, sPadded.length, ...sPadded]);

  const sequenceLen = rTlv.length + sTlv.length;
  const der = new Uint8Array(2 + sequenceLen);
  der[0] = 0x30; // SEQUENCE
  der[1] = sequenceLen;
  der.set(rTlv, 2);
  der.set(sTlv, 2 + rTlv.length);
  return der;
}

/** 去除前导零字节（保留至少一个字节） */
function trimLeadingZeros(bytes: Uint8Array): Uint8Array {
  let start = 0;
  while (start < bytes.length - 1 && bytes[start] === 0) {
    start++;
  }
  return bytes.slice(start);
}

// 导出工具函数，供其他模块使用
export { uint8ToBase64, base64ToUint8, arrayBufferToPem, pemToArrayBuffer, p1363ToDer };

File: aun/aun-sdk-core/go/crypto.go
package aun

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/pem"
	"fmt"
	"math/big"
	"time"
)

// CryptoProvider P-256 ECDSA 密钥和签名操作
type CryptoProvider struct{}

// GenerateIdentity 生成 P-256 密钥对
// 返回的 map 包含 "private_key_pem"、"public_key_der_b64"、"curve" 三个字段
func (c *CryptoProvider) GenerateIdentity() (map[string]any, error) {
	// 生成 P-256 私钥
	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("生成 P-256 密钥对失败: %w", err)
	}

	// 私钥编码为 PEM（PKCS8 格式）
	pkcs8Bytes, err := x509.MarshalPKCS8PrivateKey(privateKey)
	if err != nil {
		return nil, fmt.Errorf("序列化私钥失败: %w", err)
	}
	privateKeyPEM := pem.EncodeToMemory(&pem.Block{
		Type:  "PRIVATE KEY",
		Bytes: pkcs8Bytes,
	})

	// 公钥编码为 DER（SubjectPublicKeyInfo 格式）
	publicKeyDER, err := x509.MarshalPKIXPublicKey(&privateKey.PublicKey)
	if err != nil {
		return nil, fmt.Errorf("序列化公钥失败: %w", err)
	}

	return map[string]any{
		"private_key_pem":    string(privateKeyPEM),
		"public_key_der_b64": base64.StdEncoding.EncodeToString(publicKeyDER),
		"curve":              "P-256",
	}, nil
}

// SignLoginNonce 签名登录 nonce
// signData = "nonce:timestamp"，ECDSA SHA256 签名
// 返回 (base64 签名, 使用的时间戳, 错误)
func (c *CryptoProvider) SignLoginNonce(privateKeyPEM string, nonce string, clientTime string) (string, string, error) {
	// 解析 PEM 私钥
	block, _ := pem.Decode([]byte(privateKeyPEM))
	if block == nil {
		return "", "", fmt.Errorf("无法解析 PEM 格式私钥")
	}

	key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return "", "", fmt.Errorf("解析 PKCS8 私钥失败: %w", err)
	}

	ecKey, ok := key.(*ecdsa.PrivateKey)
	if !ok {
		return "", "", fmt.Errorf("私钥类型不是 ECDSA")
	}

	// 确定使用的时间戳
	usedTime := clientTime
	if usedTime == "" {
		usedTime = fmt.Sprintf("%f", float64(time.Now().UnixMilli())/1000.0)
	}

	// 构造签名数据
	signData := []byte(fmt.Sprintf("%s:%s", nonce, usedTime))
	hash := sha256.Sum256(signData)

	// ECDSA 签名
	r, s, err := ecdsa.Sign(rand.Reader, ecKey, hash[:])
	if err != nil {
		return "", "", fmt.Errorf("ECDSA 签名失败: %w", err)
	}

	// 编码为 ASN.1 DER 格式（与 Python cryptography 库兼容）
	sig, err := asn1MarshalECDSASignature(r, s)
	if err != nil {
		return "", "", fmt.Errorf("签名 ASN.1 编码失败: %w", err)
	}

	return base64.StdEncoding.EncodeToString(sig), usedTime, nil
}

// NewClientNonce 生成 12 字节随机 nonce（base64 编码）
func (c *CryptoProvider) NewClientNonce() string {
	nonce := make([]byte, 12)
	_, _ = rand.Read(nonce)
	return base64.StdEncoding.EncodeToString(nonce)
}

// asn1MarshalECDSASignature 将 r, s 编码为 ASN.1 DER 格式
// 与 Python cryptography 库的 ECDSA 签名格式一致
func asn1MarshalECDSASignature(r, s *big.Int) ([]byte, error) {
	rBytes := r.Bytes()
	sBytes := s.Bytes()

	// 如果最高位为 1，需要补零（ASN.1 整数规则）
	if len(rBytes) > 0 && rBytes[0]&0x80 != 0 {
		rBytes = append([]byte{0x00}, rBytes...)
	}
	if len(sBytes) > 0 && sBytes[0]&0x80 != 0 {
		sBytes = append([]byte{0x00}, sBytes...)
	}

	// ASN.1 SEQUENCE { INTEGER r, INTEGER s }
	rLen := len(rBytes)
	sLen := len(sBytes)
	totalLen := 2 + rLen + 2 + sLen // 每个 INTEGER 有 tag(1) + length(1)

	result := make([]byte, 0, 2+totalLen)
	result = append(result, 0x30)           // SEQUENCE tag
	result = append(result, byte(totalLen)) // SEQUENCE length
	result = append(result, 0x02)           // INTEGER tag
	result = append(result, byte(rLen))     // INTEGER length
	result = append(result, rBytes...)
	result = append(result, 0x02)       // INTEGER tag
	result = append(result, byte(sLen)) // INTEGER length
	result = append(result, sBytes...)

	return result, nil
}
